File: data/examples/inappropriate-intimacy.cpp
class BankAccount {
private:
    double balance;
    std::string accountNumber;
    std::string ownerName;
    
    // 允许Transaction类访问私有成员
    friend class Transaction;
    friend class AccountManager;
    
public:
    double getBalance() const { return balance; }
    std::string getAccountNumber() const { return accountNumber; }
    std::string getOwnerName() const { return ownerName; }
};

class Transaction {
private:
    BankAccount* fromAccount;
    BankAccount* toAccount;
    double amount;
    
public:
    void execute() {
        // 直接访问BankAccount的私有成员
        if (fromAccount->balance >= amount) {
            fromAccount->balance -= amount;
            toAccount->balance += amount;
        }
    }
    
    // 知道BankAccount的内部实现细节
    bool isValidAccount(BankAccount* account) {
        return account->accountNumber.length() == 10 &&
               account->balance >= 0 &&
               !account->ownerName.empty();
    }
    
    // 直接操作私有数据
    void adjustBalance(BankAccount* account, double adjustment) {
        account->balance += adjustment;
    }
};

class AccountManager {
public:
    // 也直接访问私有成员
    void freezeAccount(BankAccount* account) {
        account->balance = 0; // 直接修改私有成员
    }
    
    bool canTransfer(BankAccount* from, BankAccount* to, double amount) {
        return from->balance >= amount && 
               from->accountNumber != to->accountNumber;
    }
};

File: data/examples/inappropriate-intimacy.java
class BankAccount {
    // Package-private fields allow inappropriate access
    double balance;
    String accountNumber;
    String ownerName;
    
    public double getBalance() { return balance; }
    public String getAccountNumber() { return accountNumber; }
    public String getOwnerName() { return ownerName; }
    
    public void setBalance(double balance) { this.balance = balance; }
    public void setAccountNumber(String accountNumber) { this.accountNumber = accountNumber; }
    public void setOwnerName(String ownerName) { this.ownerName = ownerName; }
}

class Transaction {
    private BankAccount fromAccount;
    private BankAccount toAccount;
    private double amount;
    
    public void execute() {
        // Direct access to BankAccount's package-private members
        if (fromAccount.balance >= amount) {
            fromAccount.balance -= amount;
            toAccount.balance += amount;
        }
    }
    
    // Knows BankAccount's internal implementation details
    public boolean isValidAccount(BankAccount account) {
        return account.accountNumber.length() == 10 &&
               account.balance >= 0 &&
               account.ownerName != null && !account.ownerName.isEmpty();
    }
    
    // Direct manipulation of private data
    public void adjustBalance(BankAccount account, double adjustment) {
        account.balance += adjustment;
    }
    
    public void setAccounts(BankAccount from, BankAccount to) {
        this.fromAccount = from;
        this.toAccount = to;
    }
    
    public void setAmount(double amount) { this.amount = amount; }
}

class AccountManager {
    // Also directly accesses package-private members
    public void freezeAccount(BankAccount account) {
        account.balance = 0; // Direct modification of package-private member
    }
    
    public boolean canTransfer(BankAccount from, BankAccount to, double amount) {
        return from.balance >= amount && 
               !from.accountNumber.equals(to.accountNumber);
    }
}
